src/clean.c

```diff
@@ -0,0 +1,397 @@
+	}
+
+	while (true) {
+		/* Reset errno to distinguish between a previous
+		 * failure and the end of stream. */
+		errno = 0;
+		struct dirent *entry;
+		entry = readdir(dir);
+		if (!entry) {
+			if (errno) {
+				ret = errno;
+			}
+			break;
+		}
+
+		if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, "..")) {
+			continue;
+		}
+
+		char *file = NULL;
+		string_or_die(&file, "%s/%s", path, entry->d_name);
+
+		struct stat stat;
+		ret = lstat(file, &stat);
+		if (ret != 0) {
+			fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
+			goto next;
+		}
+
+		if (!options.all) {
+			if (now.tv_sec - stat.st_mtime <= seconds) {
+				goto next;
+			}
+		}
+
+		if (!pred(path, entry)) {
+			goto next;
+		}
+
+		if (options.dry_run) {
+			printf("%s\n", file);
+		} else {
+			if (S_ISDIR(stat.st_mode)) {
+				ret = rmdir(file);
+			} else {
+				ret = unlink(file);
+			}
+			if (ret != 0) {
+				fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
+			}
+		}
+		if (ret == 0) {
+			stats.files_removed++;
+		}
+	next:
+		free_string(&file);
+		ret = 0;
+	}
+
+end:
+	closedir(dir);
+	return ret;
+}
+
+static bool is_fullfile(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	return strlen(entry->d_name) == (SWUPD_HASH_LEN - 1);
+}
+
+static bool is_pack_indicator(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "pack-";
+	static const char suffix[] = ".tar";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+	static const size_t suffix_len = sizeof(suffix) - 1;
+
+	const char *name = entry->d_name;
+	size_t len = strlen(name);
+	if (len < (prefix_len + suffix_len)) {
+		return false;
+	}
+	return !memcmp(name, prefix, prefix_len) && !memcmp(name + len - suffix_len, suffix, suffix_len);
+}
+
+static bool is_all_digits(const char *s)
+{
+	while (*s) {
+		if (!isdigit(*s)) {
+			return false;
+		}
+		s++;
+	}
+	return true;
+}
+
+static bool is_manifest(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	static const char prefix[] = "Manifest.";
+	static const size_t prefix_len = sizeof(prefix) - 1;
+
+	const char *name = entry->d_name;
+	if (strlen(name) < prefix_len) {
+		return false;
+	}
+	return !memcmp(entry->d_name, prefix, prefix_len);
+}
+
+static char *read_mom_contents(int version)
+{
+	char *mom_path = NULL;
+	char *contents = NULL;
+
+	string_or_die(&mom_path, "%s/%d/Manifest.MoM", state_dir, version);
+	FILE *f = fopen(mom_path, "r");
+	if (!f) {
+		/* This is a best effort. */
+		goto end;
+	}
+
+	int ret;
+	ret = fseek(f, 0, SEEK_END);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	long size = ftell(f);
+	/* Don't care about failures or empty file. */
+	if (size < 1) {
+		goto end;
+	}
+
+	ret = fseek(f, 0, SEEK_SET);
+	if (ret != 0) {
+		goto close_then_end;
+	}
+
+	char *tmp = malloc(size + 1);
+	if (!tmp) {
+		goto close_then_end;
+	}
+
+	ret = fread(tmp, size, 1, f);
+	if (ret != 1) {
+		free_string(&tmp);
```

Comment:
should call free, not free_string. Yes we know how free_string works, and it will do the right thing, but it is breaking the abstraction. One day we might make string_or_die use alloca and make free_string just set the pointer to NULL at which time this will turn into a memory leak.

Reply:
Fixed.